### PR TITLE
GitHub Enterprise feature parity

### PR DESCRIPTION
**Pull request by jianyuan**
This Pull Request enables GitHub Enterprise integration feature parity with the GitHub integration.

### CODE REVIEW

File: src/sentry/integrations/github_enterprise/integration.py
Comment (AniketDas-Tekky) [inline src/sentry/integrations/github_enterprise/integration.py:182]: nit: update this comment with a GHE domain example

**Review by AniketDas-Tekky — COMMENTED** [review]
LGTM on the whole

**Review by AniketDas-Tekky — APPROVED** [review]

File: pyproject.toml
Comment (asottile-sentry) [inline pyproject.toml:1197]: you should be fixing errors, not extending the blocklist

**Review by asottile-sentry — DISMISSED** [review]

File: pyproject.toml
Comment (jianyuan) [reply pyproject.toml]: I have reverted the change and fixed the mypy errors.

File: tests/sentry/integrations/github_enterprise/test_client.py
Comment (asottile-sentry) [inline tests/sentry/integrations/github_enterprise/test_client.py:91]: this should instead check the type of `self.client` -- narrowing to the specific type that you're testing

File: tests/sentry/integrations/github_enterprise/test_client.py
Comment (asottile-sentry) [inline tests/sentry/integrations/github_enterprise/test_client.py:120]: it's a little weird to assert the type of a mixin -- this should assert the type of a concrete class

File: tests/sentry/integrations/github_enterprise/test_client.py
Comment (jianyuan) [reply tests/sentry/integrations/github_enterprise/test_client.py]: Fixed: I had to rename `self.client` to something else since `self.client` is already defined in `sentry.testutils.cases.TestCase`.

Comment (asottile-sentry) [reply tests/sentry/integrations/github_enterprise/test_client.py]: lg2m

Comment (jianyuan) [reply tests/sentry/integrations/github_enterprise/test_client.py]: @asottile-sentry I did some further testing. It seems that performing type narrowing in the `setUp()` method using `isinstance()` only doesn't carry the narrowed types into the test methods.

I had to use `cast()` during the assignment.

Let me know what's your preference.

Sample: https://mypy-play.net/?mypy=latest&python=3.11&flags=strict&gist=2bfe10edd440e6be3927d42e8c1c2f9e

```python
import unittest
from typing import cast


class Wide:
    pass


class Narrow(Wide):
    def hello(self) -> str:
        return "world"


def get_wide() -> Wide:
    return cast(Wide, Narrow())


# 1
class TestNarrow1(unittest.TestCase):
    def setUp(self) -> None:
        self.install = get_wide()
        assert isinstance(self.install, Narrow)

    def test_hello(self) -> None:
        # error: "Wide" has no attribute "hello"
        assert self.install.hello()


# 2
class TestNarrow2(unittest.TestCase):
    def setUp(self) -> None:
        # added cast()
        self.install = cast(Narrow, get_wide())
        assert isinstance(self.install, Narrow)

    def test_hello(self) -> None:
        # works
        assert self.install.hello()
```

Comment (asottile-sentry) [reply tests/sentry/integrations/github_enterprise/test_client.py]: yeah it would only do that in `__init__`

`cast` should really only be used for mypy bugs -- but here's a pattern that should work for you without `cast`:

```python
x = get_wide()
assert isinstance(x, Narrow)
self.x = x
```

Comment (jianyuan) [reply tests/sentry/integrations/github_enterprise/test_client.py]: Great tip! I have adjusted my PR accordingly.

File: src/sentry/integrations/github_enterprise/integration.py
Comment (leedongwei) [inline src/sentry/integrations/github_enterprise/integration.py:125]: `GitHubIntegration` extends from `CommitContextMixin` too. Github Enterprise should have commit metadata too.

Comment (leeandher) [reply src/sentry/integrations/github_enterprise/integration.py:125]: This is true, also looking at some of the procedure for CODEOWNERS, it looks like `GitHubEnterpriseIntegration` will also need the `codeowners_locations` set for it to work.

Likely also makes sense to set `repo_search` to `True`

Comment (leeandher) [reply src/sentry/integrations/github_enterprise/integration.py:125]: minor correction to my above comment* The `get_commit_context` and `CommitContextMixin` are for the suspect commits feature, not stack trace linking.

**Comment by Dhrumil-Sentry** [comment]
@leeandher When you review this PR- I think the text in the stack trace link should be the same as we have GitHub 🤔 - I don't think we need to specify GitHub Enterprise in the text

**Comment by Dhrumil-Sentry** [comment]
@leeandher In the video @jianyuan  recorded for this- Even suspect commits showed up for the issue- It would be great to know if this was the Git Blame suspect commit or the suspect commits based on associated commits. Will help with documentation and questions from SEs

**Review by leeandher — APPROVED** [review]
Looks great! Thanks for adding the tests! There are a few merge conflicts to resolve, but after that I'll trigger CI and get this merged.

@Dhrumil-Sentry per your comments:

- For the purposes of this PR I think we can leave the frontend with "Github Enterprise" since we want frontend/backend changes isolated. I also think I'm in favour of leaving it as GHE in case they have both GitHub and Github Enterprise integrations installed. Not common, but I think we always leave 'enterprise' in the title for most pages that distinguish the two (like linking/creating issues). 
image


- As far as I can tell the suspect committers are still coming from associations with commits and releases. I don't believe we use blames yet, but this may be a better question for the Issues teams, since I don't have much experience with that area

**Comment by jianyuan** [comment]
@leeandher I have fixed the merge conflicts.

**Pull request merged**
